src/nvim/ex_cmds.c

```diff
@@ -1071,6 +1071,15 @@ void do_bang(int addr_count, exarg_T *eap, bool forceit, bool do_in, bool do_out
   } else {                            // :range!
     // Careful: This may recursively call do_bang() again! (because of
     // autocommands)
+    if (strlen(newcmd) == 0 && p_sh != NULL) {
```

Comment:
a small hint is high-leverage
```suggestion
    if (strlen(newcmd) == 0 && p_sh != NULL) {  // ":range!" without command pipes to 'shell'.
```

Reply:
Agreed, this is done.